src/sentry/api/endpoints/user_authenticator_enroll.py

```diff
@@ -0,0 +1,202 @@
+from sentry.api.decorators import sudo_required
+from sentry.api.serializers import serialize
+from sentry.models import Authenticator
+from sentry.security import capture_security_activity
+
+
+class TotpRestSerializer(serializers.Serializer):
+    otp = serializers.CharField(
+        label='One-time password',
+        help_text='Code from authenticator',
+        required=True,
+        max_length=20
+    )
+
+
+class SmsRestSerializer(serializers.Serializer):
+    phone = serializers.CharField(
+        label="Phone number",
+        help_text="Phone number to send SMS code",
+        required=True,
+        max_length=20,
+    )
+    otp = serializers.CharField(
+        label='One-time password',
+        help_text='Code from authenticator',
+        required=False,
+        max_length=20
+    )
+
+
+class U2fRestSerializer(serializers.Serializer):
+    deviceName = serializers.CharField(
+        label='Device name',
+        required=False,
+        max_length=60,
+        default=lambda: petname.Generate(2, ' ', letters=10).title(),
+    )
+
+serializer_map = {
+    'totp': TotpRestSerializer,
+    'sms': SmsRestSerializer,
+    'u2f': U2fRestSerializer,
+}
+
+
+def get_serializer_field_metadata(serializer, fields=None):
+    """Returns field metadata for serializer"""
+    meta = []
+    for field in serializer.base_fields:
+        if fields is None or field in fields:
+            serialized_field = dict(serializer.base_fields[field].metadata())
+            serialized_field['name'] = field
+            serialized_field['defaultValue'] = serializer.base_fields[field].get_default_value()
+            meta.append(serialized_field)
+
+    return meta
+
+
+class UserAuthenticatorEnrollEndpoint(UserEndpoint):
+    permission_classes = (IsAuthenticated, UserPermission, )
+
+    # @sudo_required
+    def get(self, request, user, interface_id):
+        """
+        Get Authenticator Interface
+        ```````````````````````````
+
+        Retrieves authenticator interface details for user depending on user enrollment status
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        interface = Authenticator.objects.get_interface(user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        # User is not enrolled in auth interface:
+        # - display configuration form
+        response = serialize(interface)
+        response['form'] = get_serializer_field_metadata(
+            serializer_map[interface_id]
+        )
+
+        # U2fInterface has no 'secret' attribute
+        try:
+            response['secret'] = interface.secret
+        except AttributeError:
+            pass
+
+        if interface_id == 'totp':
```

Comment:
More serialization based on `interface_id`, also in this endpoint, `interface.is_enrolled` should be `false` (or allow multiple enrollments e.g. u2f).